권한별 Drawer 메뉴 관련 로직 처리

Na2te opened the pull request:
## 📌 연결된 이슈
- #140 

### ✅ 작업 내용
- 내 스터디 목록 보는 API 연동
- 권한별 Drawer 삭제 메뉴 분기처리
- 권한별 그룹 수정 화면 분기처리
- 권한별 그룹원 목록 화면 분기처리
- 권한별 그룹 가입 신청 목록 화면 분기처리
- 삭제 시 Dialog 나오도록 처리

### 📷 관련 이미지(영상)
+ #### Drawer 화면

  리더 외엔 그룹 삭제 권한이 없으므로 Drawer 메뉴에서 안보이게 처리     

|리더|리더 외|
|:--:|:--:|
|![image](https://github.com/user-attachments/assets/852392c7-866b-477b-ad29-5be17d1b5408)|![image](https://github.com/user-attachments/assets/d0fb856a-3257-4dae-b67c-98d3594c1614)|
|<img width="221" src="https://github.com/user-attachments/assets/b92c4484-38d2-475f-8f35-810790c60053" />||
|<img width="221" src="https://github.com/user-attachments/assets/ad4de0b4-e13b-4686-9f12-9a3911af3f4d" />|<img width="221" src="https://github.com/user-attachments/assets/ad4de0b4-e13b-4686-9f12-9a3911af3f4d" />|

+ #### 그룹 정보 수정 화면

  리더 외엔 수정 권한이 없으므로 리더 외에는 수정할 수 있는 정보를 전부 disable처리 및 수정 버튼 안보이게 처리

|리더|리더 외|
|:--:|:--:|
|![image](https://github.com/user-attachments/assets/885fa5c6-9542-48aa-a231-0231cec9010f)|![image](https://github.com/user-attachments/assets/16eae41d-8af9-4b9d-ae82-847dec09bb95)|


+ #### 그룹원 목록 및 직책 변경 화면

  리더 외엔 수정 권한이 없으므로 리더 외에는 직책, 강퇴 버튼 안보이게 처리

|리더|리더 외|
|:--:|:--:|
|![image](https://github.com/user-attachments/assets/4f9a2fae-2836-49e8-bbc3-bf7b64d9aac1)|![image](https://github.com/user-attachments/assets/2493124e-74c6-4acf-a2c2-d75ab8f10770)|

+ #### 그룹 가입 신청 목록 화면

  리더, 매니저 외엔 수정 권한이 없으므로 리더, 매니저 외에는 수락 거절 버튼 안보이게 처리

|리더, 매니저|리더, 매니저 외|
|:--:|:--:|
|![image](https://github.com/user-attachments/assets/e7fbf882-8856-448b-9436-6f4701f45c54)|![image](https://github.com/user-attachments/assets/4cf80bf5-a592-4e0f-af77-02b7dcf9167c)|

### 🤔 의논하고 싶은 내용(코드 리뷰 받고 싶은 부분)
- <mark>우선 Drawer의 내부 변수로 Role을 Leader로 설정해놓았습니다, 이건 나중에 Merge 되고 스터디 상세 조회 API 연결되면서 state 내부에 role이 생기면 그걸 사용하는 거로 수정해야 합니다</mark>
- 삭제 다이얼로그가 폰트가 설정되어 있지 않긴 한데, 이건 나중에 MVP 출시 후 기획 다시 정리하면서 좀 맞춰야 하는 부분 아닐까 생각하고 있는데 혹 아니라면 저거 폰트 어떤거로 설정하는게 좋을까요?
- 내 스터디 클릭 시 해당 스터디로 이동하는 기능 및 더보기로 스터디 목록들 보는 화면은 중요 사안은 아니라고 보여져서 1차 MVP 뒤로 미뤘습니다
- 확실하게 1차 MVP 뒤로 뺄거면 더보기 메뉴는 없애야 할 거 같긴 합니다 없앨까요??

Na2te committed:
fix(#140): 불필요한 Drawer 메뉴 삭제 및 Lazy Column 처리
- 그룹 생성, 그룹 검색, 일정 관리 메뉴 삭제
- Height가 작은 휴대폰의 경우 Column은 자동으로 스크롤을 채우지 않아 잘리던 문제 해결
- Lazy Column으로 영역 넘어갈 시 Scroll 되도록 함
- 메뉴 삭제로 안쓰는 인자 삭제

todayis-sunny committed:
refactor(#143): DrawerState Content 영역에 선언 (프리뷰)

- compositionLocal이 들어가 있으면 미리보기 불가

Na2te committed:
feat(#140): 스터디 그룹 상세 조회 API 연동
- 그룹 상세를 조회하는 Api, DataSource, Repository, Usecase 생성 완료

Na2te committed:
feat(#140): 내 스터디 목록 조회 API 연동
- Paging3로 관련 작업 및 UseCase 생성 완료

Na2te committed:
feat(#140): Drawer에 내 스터디 그룹 조회 적용
- 스터디 그룹 내에서 열 수 있는 Drawer 내에 내 스터디 그룹 조회 적용
- 현재 있는 스터디 그룹을 따로 제외하지 않음

Na2te committed:
feat(#140): 역할 모델 생성
- 스터디에서 자신의 위치를 구분하는 Role Enum Class 생성

Na2te committed:
fix(#140): Drawer 로직 수정
- Navigate시 권한 정보를 넘기는 식으로 람다를 수정함
- Drawer에서 삭제 메뉴는 리더만 볼 수 있도록 설정

Na2te committed:
feat(#140): 스터디원 신청 목록 페이지 로직 추가
- Manager, Leader가 아니면 신청 버튼 숨김처리

Na2te committed:
feat(#140): 멤버 관리 로직 추가
- Leader가 아니면 직책 변경 메뉴가 보이지 않도록 수정

Na2te committed:
feat(#140): 그룹 정보 수정 기능 로직 추가
- 리더가 아니면 수정이 막히도록 설정

Na2te committed:
style(#140): Preview 추가
- Leader일 때 Drawer의 Preview 화면 추가

Na2te committed:
fix(#140): Drawer 내 스터디 보기 정렬 기준 수정
- spaceBy로 앞에서부터 채우면서 간격 설정

Na2te committed:
feat(#140): 삭제 다이얼로그 추가
- 그룹 삭제 누를 시 확인 다이얼로그가 나오도록 설정

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/data/paging/MyStudyGroupListPagingSource.kt:
```diff
@@ -0,0 +1,48 @@
+import com.ssafy.neegongnaegong.domain.model.studygroup.MyStudyGroupInfo
+import kotlinx.coroutines.flow.first
+
+class MyStudyGroupListPagingSource(
+    private val dataSource: NetworkStudyGroupDataSource,
+) : PagingSource<CursorSliceKey, MyStudyGroupInfo>() {
+    override suspend fun load(params: LoadParams<CursorSliceKey>): LoadResult<CursorSliceKey, MyStudyGroupInfo> {
+        return try {
+            val cursor = params.key
+
+            val response =
+                with(cursor) {
+                    dataSource.getMyStudyGroupList(
+                        this?.cursorCreatedAt,
+                        this?.cursorId,
+                        10,
+                    ).first()
+                }
+
+            val data: List<MyStudyGroupListResponse.MyStudyGroupResponse> = response.content
+            LoadResult.Page(
+                data = data.map { it.toDomain() },
+                prevKey = null,
+                nextKey =
+                    if (response.hasNext) {
+                        CursorSliceKey(response.cursorCreatedAt, response.cursorId)
+                    } else {
+                        null
+                    },
+            )
+        } catch (e: Exception) {
+            LoadResult.Error(e)
+        }
+    }
+
+    override fun getRefreshKey(state: PagingState<CursorSliceKey, MyStudyGroupInfo>): CursorSliceKey? {
+        return null
+    }
+}
```
`paging` 잘 배워갑니다!

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/domain/model/studygroup/MyStudyGroupInfo.kt:
```diff
@@ -0,0 +1,33 @@
+
+import java.time.LocalDate
+
+data class MyStudyGroupInfo(
+    val id: Long,
+    val leader: LeaderInfo,
+    val name: String,
+    val maxMembers: Int,
+    val currentMembers: Int,
+    val description: String,
+    val profileImg: String,
+    val isPublic: Boolean,
+    val targetStudyTime: Int,
+    val category: CategoryInfo,
+    val createdDate: LocalDate,
+    val tags: List<TagInfo>,
+) {
+    data class LeaderInfo(
+        val id: Long,
+        val name: String,
+    )
+
+    data class CategoryInfo(
+        val id: Long,
+        val name: String,
+    )
+
+    data class TagInfo(
+        val id: Long,
+        val name: String,
+    )
+}
```
`Nested Class` 패턴으로 적용하니 좋습니다:)

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/component/drawer/model/Role.kt:
```diff
@@ -0,0 +1,10 @@
+package com.ssafy.neegongnaegong.presentation.group.component.drawer.model
+
+const val TEAM_LEADER = "TEAM_LEADER"
```
이 상수는 어디서 사용되는 걸까요? 🤔

Na2te replied:
앗...
지우는 걸 깜박한 것 같습니다 수정하겠습니다

Na2te replied:
수정했습니다
[afbbcb4](https://github.com/NeeGongNaeGong/android/pull/167/commits/afbbcb49f499c2a6f219ca7dfe9a6a1f8940d3ee) 커밋 확인해주시면 될 것 같습니다!

todayis-sunny reviewed (approved):
> 우선 Drawer의 내부 변수로 Role을 Leader로 설정해놓았습니다, 이건 나중에 Merge 되고 스터디 상세 조회 API 연결되면서 state 내부에 role이 생기면 그걸 사용하는 거로 수정해야 합니다

확인했습니다.

> 삭제 다이얼로그가 폰트가 설정되어 있지 않긴 한데, 이건 나중에 MVP 출시 후 기획 다시 정리하면서 좀 맞춰야 하는 부분 아닐까 생각하고 있는데 혹 아니라면 저거 폰트 어떤거로 설정하는게 좋을까요?

제가 폰트를 기입하지 못한것 같습니다. **우리 앱의 테마** 스타일 대로 작성하면 좋을것 같습니다!

> 확실하게 1차 MVP 뒤로 뺄거면 더보기 메뉴는 없애야 할 거 같긴 합니다 없앨까요??

더보기 메뉴는 없애도 좋을거 같습니다!

드로어 쪽 관련해서 작업하기 복잡했을텐데, 잘 작업해주신것 같습니다.
고생하셨습니다!

Na2te committed:
refactor(#140): 불필요한 상수 제거
- enum으로 처리해 쓰지 않는 LEADER 상수 제거

Na2te committed:
fix(#140): 내 스터디 목록 NextKey 로직 수정
- 스터디 정보마다 cursor 속성 추가 및 mapper 설정
- 맨 앞 아이템의 cursor 값을 이용하도록 변경
- Preview에 필요한 인자 추가

Na2te committed:
fix(#140): 내 스터디 목록 더보기 메뉴 삭제
- MVP에 제외된 기능이라 삭제

Na2te committed:
fix(#140): 자동 검사 로직 수정
- 기존에는 add 되는 파일들에 대해 detekt를 돌리고 검사하는 로직
- 이 경우 파일의 일부만 add 해서 추가할 경우에도 코드 자동 정렬 이후 같이 add 되는 문제 존재
- 이를 해결

Na2te commented:
> > 삭제 다이얼로그가 폰트가 설정되어 있지 않긴 한데, 이건 나중에 MVP 출시 후 기획 다시 정리하면서 좀 맞춰야 하는 부분 아닐까 생각하고 있는데 혹 아니라면 저거 폰트 어떤거로 설정하는게 좋을까요?
> 
> 제가 폰트를 기입하지 못한것 같습니다. **우리 앱의 테마** 스타일 대로 작성하면 좋을것 같습니다!
> 
Figma에 디자인이 없어서 정확하게 무슨 폰트를 지정할 지 애매해서 나중에 기획 다시 정리할 때 없는 디자인 Figma에 작성하면서 정확하게 결정나면 그대로 하면 될 것 같습니다! 지금 수정해도 결국 나중에 수정해야 할 것 같아서 작업 2번하는 게 될 것 같습니다

> > 확실하게 1차 MVP 뒤로 뺄거면 더보기 메뉴는 없애야 할 거 같긴 합니다 없앨까요??
> 
> 더보기 메뉴는 없애도 좋을거 같습니다!
> 
수정했습니다
[12b9113](https://github.com/NeeGongNaeGong/android/pull/167/commits/12b91133e94d16708e3bedc50f746708d951787e) 커밋 확인해주시면 될 것 같습니다!

Na2te commented:
추가적으로 기존에 파일의 일부분만 add를 해서 commit을 반영하려고 해도 commit 시 코드 자동 검사 후 add를 하는 과정에서 파일 단위로 add가 되어 add 하지 않고자 했던 부분도 같이 add 되어 버리는 이슈를 발견했습니다
이를 해결하고자 [7d839af](https://github.com/NeeGongNaeGong/android/pull/167/commits/7d839afdd101eee128c49c2d77a71922603cb57a) 커밋으로 해결하였으니 참고 바랍니다

todayis-sunny commented:
수정사항 다 확인했습니다.
고생하셧습니다 :)
